mopy/core/spectrumgroup.py

```diff
@@ -364,61 +564,171 @@ def _calc_spectral_params(self):
         # Get moment
-        moment = source_df["omega0"] * 4 * np.pi * source_velocity ** 3 * density
+        moment = omega0 * 4 * np.pi * source_velocity ** 3 * density
         moment.name = "moment"
-        # Get moment magnitude
-        mw = (2 / 3) * np.log10(moment) - 6.0
+        return moment
+
+    def calc_moment_mag(self, moment: pd.Series) -> pd.Series:
+        """
+        Calculate moment magnitude
+
+        Parameters
+        ----------
+        moment
+            Seismic moment (N-m) calculated for each station/phase/event
+        """
+        mw = 2.0 / 3.0 * (np.log10(moment) - 9.1)
         mw.name = "mw"
-        # Get potency
-        potency = source_df["omega0"] * 4 * np.pi * source_velocity
+        return mw
+
+    def calc_potency(self, omega0: pd.Series) -> pd.Series:
+        """
+        Calculate the seismic potency
+
+        Parameters
+        ----------
+        omega0
+            List of omega0 calculated for each trace
+
+        Returns
+        -------
+        Seismic potency in m^2/s^2. All components of a station will be combined for each event/pick.
+        """
+        source_velocity = (
+            self.stats["source_velocity"]
+            .groupby(["phase_hint", "event_id", "seed_id_less"])
+            .mean()
+        )  # TODO: This should work because this should be the same across all three components (barring anisotropy or something equally weird)... this is definitely a bandaid to solve my particular problem, though
+        potency = (
+            omega0 * 4 * np.pi * source_velocity
+        )  # The other parameters should have already been corrected
         potency.name = "potency"
-        # Get energy
-        energy = 4 * np.pi * source_df["velocity_squared_integral"] * density * source_velocity
+        return potency
+
+    def calc_energy(self) -> pd.Series:
+        """
+        Calculate the radiated seismic energy
+
+        Returns
+        -------
+        Radiated seismic energy in Joules. All components of a station will be combined for each event/pick.
+        """
+        sg = self.abs()
+        # Get the velocity psd and integrate
+        vel_psd = sg.to_spectra_type("psd", motion_type="velocity")
+        vel_psd_per_station = vel_psd.data.groupby(
+            ["phase_hint", "event_id", "seed_id_less"]
+        ).apply(
+            np.linalg.norm
+        )  # TODO: Or should this be a straight summation? Also, should this be done before getting the PSD?
+        vel_psd_int = vel_psd_per_station.sum()
+        # Get necessary values from the stats dataframe
+        density = (
+            self.stats["density"]
+            .groupby(["phase_hint", "event_id", "seed_id_less"])
+            .mean()
+        )  # TODO: This should work because this should be the same across all three components (barring anisotropy or something equally weird)... this is definitely a bandaid to solve my particular problem, though
```

Comment:
I think this should be fine for both density and velocity since we are grouping by phase and station